File: frontend/src/utils/addressUtils.js
/**
 * WEPO Address Standardization System
 * Unified address generation, validation, and formatting
 */

import CryptoJS from 'crypto-js';
import * as bitcoin from 'bitcoinjs-lib';
import ECPairFactory from 'ecpair';
import * as ecc from 'tiny-secp256k1';

// Initialize ECPair with secp256k1 implementation
const ECPair = ECPairFactory(ecc);

// Address format constants
export const ADDRESS_FORMATS = {
  WEPO_REGULAR: {
    prefix: 'wepo1',
    payloadLength: 32,
    totalLength: 37, // wepo1 + 32 chars
    type: 'regular'
  },
  WEPO_QUANTUM: {
    prefix: 'wepo1q',
    payloadLength: 39,
    totalLength: 45, // wepo1q + 39 chars  
    type: 'quantum'
  },
  BTC_LEGACY: {
    prefix: '1',
    minLength: 26,
    maxLength: 35,
    type: 'bitcoin-legacy'
  },
  BTC_SEGWIT: {
    prefix: 'bc1',
    minLength: 14,
    maxLength: 74,
    type: 'bitcoin-segwit'
  }
};

/**
 * Generate Bitcoin address from seed
 * @param {string|Buffer} seed - Wallet seed
 * @param {string} addressType - 'legacy' or 'segwit'
 * @returns {object} Bitcoin address and private key
 */
export const generateBitcoinAddress = (seed, addressType = 'legacy') => {
  try {
    const seedString = typeof seed === 'string' ? seed : seed.toString('hex');
    
    // Create a deterministic private key from seed
    const hash = CryptoJS.SHA256(seedString + 'bitcoin').toString();
    const privateKeyHex = hash.substring(0, 64);
    
    // Convert to Buffer for bitcoinjs-lib
    const privateKeyBuffer = Buffer.from(privateKeyHex, 'hex');
    
    // Create key pair
    const keyPair = bitcoin.ECPair.fromPrivateKey(privateKeyBuffer);
    
    let address;
    if (addressType === 'segwit') {
      // Generate P2WPKH (native segwit) address
      const { address: segwitAddress } = bitcoin.payments.p2wpkh({ 
        pubkey: keyPair.publicKey 
      });
      address = segwitAddress;
    } else {
      // Generate P2PKH (legacy) address
      const { address: legacyAddress } = bitcoin.payments.p2pkh({ 
        pubkey: keyPair.publicKey 
      });
      address = legacyAddress;
    }
    
    return {
      address,
      privateKey: privateKeyHex,
      publicKey: keyPair.publicKey.toString('hex'),
      type: addressType
    };
  } catch (error) {
    console.error('Bitcoin address generation error:', error);
    // Fallback to simple hash-based address for testing
    const seedString = typeof seed === 'string' ? seed : seed.toString('hex');
    const hash = CryptoJS.SHA256(seedString + 'bitcoin').toString();
    const simpleAddress = '1' + hash.substring(0, 33); // Simple 34-char address
    
    return {
      address: simpleAddress,
      privateKey: hash.substring(0, 64),
      publicKey: hash.substring(0, 66),
      type: 'simple'
    };
  }
};
/**
 * Generate standardized WEPO address from seed
 * @param {string|Buffer} seed - Wallet seed
 * @param {string} type - 'regular' or 'quantum'
 * @returns {string} Standardized WEPO address
 */
export const generateWepoAddress = (seed, type = 'regular') => {
  const seedString = typeof seed === 'string' ? seed : seed.toString('hex');
  const hash = CryptoJS.SHA256(seedString).toString();
  
  if (type === 'quantum') {
    const format = ADDRESS_FORMATS.WEPO_QUANTUM;
    const payload = hash.substring(0, format.payloadLength);
    return `${format.prefix}${payload}`;
  } else {
    const format = ADDRESS_FORMATS.WEPO_REGULAR;
    const payload = hash.substring(0, format.payloadLength);
    return `${format.prefix}${payload}`;
  }
};

/**
 * Validate address format (WEPO or Bitcoin)
 * @param {string} address - Address to validate
 * @returns {object} Validation result with type detection
 */
export const validateAddress = (address) => {
  if (!address || typeof address !== 'string') {
    return { 
      valid: false, 
      type: null, 
      error: 'Address must be a string' 
    };
  }

  // Check for Bitcoin addresses first
  const btcValidation = validateBitcoinAddress(address);
  if (btcValidation.valid) {
    return btcValidation;
  }

  // Check for WEPO addresses
  return validateWepoAddress(address);
};

/**
 * Validate Bitcoin address format
 * @param {string} address - Bitcoin address to validate
 * @returns {object} Validation result
 */
export const validateBitcoinAddress = (address) => {
  if (!address || typeof address !== 'string') {
    return { 
      valid: false, 
      type: null, 
      error: 'Address must be a string' 
    };
  }

  // Legacy Bitcoin address (P2PKH)
  if (address.startsWith('1')) {
    const format = ADDRESS_FORMATS.BTC_LEGACY;
    if (address.length >= format.minLength && address.length <= format.maxLength) {
      return { 
        valid: true, 
        type: 'bitcoin-legacy', 
        format: format 
      };
    }
  }

  // Segwit Bitcoin address (P2WPKH)
  if (address.startsWith('bc1')) {
    const format = ADDRESS_FORMATS.BTC_SEGWIT;
    if (address.length >= format.minLength && address.length <= format.maxLength) {
      return { 
        valid: true, 
        type: 'bitcoin-segwit', 
        format: format 
      };
    }
  }

  // Multisig Bitcoin address (P2SH)
  if (address.startsWith('3')) {
    if (address.length >= 26 && address.length <= 35) {
      return { 
        valid: true, 
        type: 'bitcoin-multisig', 
        format: { prefix: '3', type: 'bitcoin-multisig' }
      };
    }
  }

  return { 
    valid: false, 
    type: null, 
    error: 'Invalid Bitcoin address format' 
  };
};
/**
 * Validate WEPO address format
 * @param {string} address - Address to validate
 * @returns {object} Validation result with type detection
 */
export const validateWepoAddress = (address) => {
  if (!address || typeof address !== 'string') {
    return { 
      valid: false, 
      type: null, 
      error: 'Address must be a string' 
    };
  }

  // Check for regular WEPO address
  if (address.startsWith(ADDRESS_FORMATS.WEPO_REGULAR.prefix) && 
      !address.startsWith(ADDRESS_FORMATS.WEPO_QUANTUM.prefix)) {
    
    if (address.length === ADDRESS_FORMATS.WEPO_REGULAR.totalLength) {
      return { 
        valid: true, 
        type: 'regular', 
        format: ADDRESS_FORMATS.WEPO_REGULAR 
      };
    } else {
      return { 
        valid: false, 
        type: 'regular', 
        error: `Regular WEPO address must be ${ADDRESS_FORMATS.WEPO_REGULAR.totalLength} characters` 
      };
    }
  }

  // Check for quantum WEPO address
  if (address.startsWith(ADDRESS_FORMATS.WEPO_QUANTUM.prefix)) {
    if (address.length === ADDRESS_FORMATS.WEPO_QUANTUM.totalLength) {
      return { 
        valid: true, 
        type: 'quantum', 
        format: ADDRESS_FORMATS.WEPO_QUANTUM 
      };
    } else {
      return { 
        valid: false, 
        type: 'quantum', 
        error: `Quantum WEPO address must be ${ADDRESS_FORMATS.WEPO_QUANTUM.totalLength} characters` 
      };
    }
  }

  return { 
    valid: false, 
    type: null, 
    error: 'Address must start with wepo1 (regular) or wepo1q (quantum)' 
  };
};

/**
 * Detect address type from address string
 * @param {string} address - Address to analyze
 * @returns {string|null} Address type or null if invalid
 */
export const getAddressType = (address) => {
  const validation = validateAddress(address);
  return validation.valid ? validation.type : null;
};

/**
 * Check if address is Bitcoin
 * @param {string} address - Address to check
 * @returns {boolean} True if Bitcoin address
 */
export const isBitcoinAddress = (address) => {
  const type = getAddressType(address);
  return type && type.startsWith('bitcoin');
};

/**
 * Check if address is WEPO
 * @param {string} address - Address to check
 * @returns {boolean} True if WEPO address
 */
export const isWepoAddress = (address) => {
  const type = getAddressType(address);
  return type === 'regular' || type === 'quantum';
};

/**
 * Check if address is quantum-resistant
 * @param {string} address - Address to check
 * @returns {boolean} True if quantum-resistant
 */
export const isQuantumAddress = (address) => {
  return getAddressType(address) === 'quantum';
};

/**
 * Format address for display (truncate middle)
 * @param {string} address - Full address
 * @param {number} startChars - Characters to show at start
 * @param {number} endChars - Characters to show at end
 * @returns {string} Formatted address
 */
export const formatAddressForDisplay = (address, startChars = 8, endChars = 6) => {
  if (!address || address.length <= startChars + endChars) {
    return address;
  }
  
  return `${address.substring(0, startChars)}...${address.substring(address.length - endChars)}`;
};

/**
 * Convert legacy address to standardized format
 * @param {string} legacyAddress - Old format address
 * @returns {string} Standardized address
 */
export const standardizeAddress = (legacyAddress) => {
  if (!legacyAddress || !legacyAddress.startsWith('wepo1')) {
    return legacyAddress;
  }

  // Handle legacy 45-char addresses (convert to quantum format)
  if (legacyAddress.length === 45 && !legacyAddress.startsWith('wepo1q')) {
    // Convert to new quantum format
    const payload = legacyAddress.substring(5, 44); // Remove 'wepo1', take 39 chars
    return `wepo1q${payload}`;
  }

  // Handle legacy 37-char addresses (already standard regular format)
  if (legacyAddress.length === 37) {
    return legacyAddress; // Already in correct format
  }

  // Return as-is if doesn't match expected patterns
  return legacyAddress;
};

/**
 * Generate address validation regex patterns
 * @returns {object} Regex patterns for different address types
 */
export const getAddressPatterns = () => {
  return {
    regular: new RegExp(`^${ADDRESS_FORMATS.WEPO_REGULAR.prefix}[a-f0-9]{${ADDRESS_FORMATS.WEPO_REGULAR.payloadLength}}$`),
    quantum: new RegExp(`^${ADDRESS_FORMATS.WEPO_QUANTUM.prefix}[a-f0-9]{${ADDRESS_FORMATS.WEPO_QUANTUM.payloadLength}}$`),
    any: new RegExp(`^wepo1q?[a-f0-9]{32,39}$`)
  };
};

/**
 * Check if two addresses are equivalent (handles legacy formats)
 * @param {string} address1 - First address
 * @param {string} address2 - Second address
 * @returns {boolean} True if addresses are equivalent
 */
export const addressesEqual = (address1, address2) => {
  if (!address1 || !address2) return false;
  
  const std1 = standardizeAddress(address1);
  const std2 = standardizeAddress(address2);
  
  return std1 === std2;
};

/**
 * Batch validate multiple addresses
 * @param {string[]} addresses - Array of addresses to validate
 * @returns {object[]} Array of validation results
 */
export const validateAddressBatch = (addresses) => {
  return addresses.map(address => ({
    address,
    ...validateWepoAddress(address)
  }));
};

export default {
  ADDRESS_FORMATS,
  generateWepoAddress,
  generateBitcoinAddress,
  validateAddress,
  validateWepoAddress,
  validateBitcoinAddress,
  getAddressType,
  isBitcoinAddress,
  isWepoAddress,
  isQuantumAddress,
  formatAddressForDisplay,
  standardizeAddress,
  getAddressPatterns,
  addressesEqual,
  validateAddressBatch
};

File: wepo-blockchain/core/address_utils.py
#!/usr/bin/env python3
"""
WEPO Address Standardization System - Backend
Unified address generation, validation, and formatting for Python backend
"""

import hashlib
import re
from typing import Dict, Optional, Tuple, Union

# Address format definitions
ADDRESS_FORMATS = {
    "WEPO_REGULAR": {
        "prefix": "wepo1",
        "payload_length": 32,
        "total_length": 37,
        "type": "regular"
    },
    "WEPO_QUANTUM": {
        "prefix": "wepo1q",
        "payload_length": 39,
        "total_length": 45,
        "type": "quantum"
    },
    "BTC": {
        "prefix": "1",
        "payload_length": 25,
        "total_length": 34,
        "type": "bitcoin"
    }
}

def generate_wepo_address(seed: Union[str, bytes], address_type: str = "regular") -> str:
    """
    Generate standardized WEPO address from seed
    
    Args:
        seed: Wallet seed (string or bytes)
        address_type: 'regular' or 'quantum'
        
    Returns:
        Standardized WEPO address
    """
    if isinstance(seed, bytes):
        seed_str = seed.hex()
    else:
        seed_str = str(seed)
    
    # Generate hash from seed
    hash_obj = hashlib.sha256(seed_str.encode()).hexdigest()
    
    if address_type == "quantum":
        format_info = ADDRESS_FORMATS["WEPO_QUANTUM"]
        payload = hash_obj[:format_info["payload_length"]]
        return f"{format_info['prefix']}{payload}"
    else:
        format_info = ADDRESS_FORMATS["WEPO_REGULAR"]
        payload = hash_obj[:format_info["payload_length"]]
        return f"{format_info['prefix']}{payload}"

def validate_wepo_address(address: str) -> Dict[str, Union[bool, str, None]]:
    """
    Validate WEPO address format
    
    Args:
        address: Address string to validate
        
    Returns:
        Dict with validation result, type, and error message if invalid
    """
    if not address or not isinstance(address, str):
        return {
            "valid": False,
            "type": None,
            "error": "Address must be a string"
        }
    
    # Check for regular WEPO address
    regular_format = ADDRESS_FORMATS["WEPO_REGULAR"]
    if address.startswith(regular_format["prefix"]) and not address.startswith(ADDRESS_FORMATS["WEPO_QUANTUM"]["prefix"]):
        if len(address) == regular_format["total_length"]:
            return {
                "valid": True,
                "type": "regular",
                "format": regular_format
            }
        else:
            return {
                "valid": False,
                "type": "regular",
                "error": f"Regular WEPO address must be {regular_format['total_length']} characters"
            }
    
    # Check for quantum WEPO address
    quantum_format = ADDRESS_FORMATS["WEPO_QUANTUM"]
    if address.startswith(quantum_format["prefix"]):
        if len(address) == quantum_format["total_length"]:
            return {
                "valid": True,
                "type": "quantum",
                "format": quantum_format
            }
        else:
            return {
                "valid": False,
                "type": "quantum",
                "error": f"Quantum WEPO address must be {quantum_format['total_length']} characters"
            }
    
    return {
        "valid": False,
        "type": None,
        "error": "Address must start with wepo1 (regular) or wepo1q (quantum)"
    }

def get_address_type(address: str) -> Optional[str]:
    """
    Detect address type from address string
    
    Args:
        address: Address to analyze
        
    Returns:
        Address type ('regular', 'quantum') or None if invalid
    """
    validation = validate_wepo_address(address)
    return validation["type"] if validation["valid"] else None

def is_quantum_address(address: str) -> bool:
    """
    Check if address is quantum-resistant
    
    Args:
        address: Address to check
        
    Returns:
        True if quantum-resistant
    """
    return get_address_type(address) == "quantum"

def is_regular_address(address: str) -> bool:
    """
    Check if address is regular WEPO address
    
    Args:
        address: Address to check
        
    Returns:
        True if regular WEPO address
    """
    return get_address_type(address) == "regular"

def format_address_for_display(address: str, start_chars: int = 8, end_chars: int = 6) -> str:
    """
    Format address for display (truncate middle)
    
    Args:
        address: Full address
        start_chars: Characters to show at start
        end_chars: Characters to show at end
        
    Returns:
        Formatted address
    """
    if not address or len(address) <= start_chars + end_chars:
        return address
    
    return f"{address[:start_chars]}...{address[-end_chars:]}"

def standardize_address(legacy_address: str) -> str:
    """
    Convert legacy address to standardized format
    
    Args:
        legacy_address: Old format address
        
    Returns:
        Standardized address
    """
    if not legacy_address or not legacy_address.startswith("wepo1"):
        return legacy_address
    
    # Handle legacy 45-char addresses (convert to quantum format)
    if len(legacy_address) == 45 and not legacy_address.startswith("wepo1q"):
        # Convert to new quantum format
        payload = legacy_address[5:44]  # Remove 'wepo1', take 39 chars
        return f"wepo1q{payload}"
    
    # Handle legacy 37-char addresses (already standard regular format)
    if len(legacy_address) == 37:
        return legacy_address  # Already in correct format
    
    # Return as-is if doesn't match expected patterns
    return legacy_address

def get_address_patterns() -> Dict[str, re.Pattern]:
    """
    Generate address validation regex patterns
    
    Returns:
        Dict of regex patterns for different address types
    """
    regular_format = ADDRESS_FORMATS["WEPO_REGULAR"]
    quantum_format = ADDRESS_FORMATS["WEPO_QUANTUM"]
    
    return {
        "regular": re.compile(f"^{re.escape(regular_format['prefix'])}[a-f0-9]{{{regular_format['payload_length']}}}$"),
        "quantum": re.compile(f"^{re.escape(quantum_format['prefix'])}[a-f0-9]{{{quantum_format['payload_length']}}}$"),
        "any": re.compile(r"^wepo1q?[a-f0-9]{32,39}$")
    }

def addresses_equal(address1: str, address2: str) -> bool:
    """
    Check if two addresses are equivalent (handles legacy formats)
    
    Args:
        address1: First address
        address2: Second address
        
    Returns:
        True if addresses are equivalent
    """
    if not address1 or not address2:
        return False
    
    std1 = standardize_address(address1)
    std2 = standardize_address(address2)
    
    return std1 == std2

def validate_address_batch(addresses: list) -> list:
    """
    Batch validate multiple addresses
    
    Args:
        addresses: List of addresses to validate
        
    Returns:
        List of validation results
    """
    results = []
    for address in addresses:
        validation = validate_wepo_address(address)
        results.append({
            "address": address,
            **validation
        })
    return results

def is_valid_wepo_address(address: str) -> bool:
    """
    Simple boolean check for address validity
    
    Args:
        address: Address to validate
        
    Returns:
        True if valid WEPO address
    """
    validation = validate_wepo_address(address)
    return validation["valid"]

# Legacy compatibility functions
def validate_btc_address(address: str) -> bool:
    """
    Legacy BTC address validation (placeholder for future BTC integration)
    
    Args:
        address: BTC address to validate
        
    Returns:
        True if valid BTC address format
    """
    # Simple validation for now - will be enhanced with proper BTC integration
    if not address or not isinstance(address, str):
        return False
    
    # Basic BTC address patterns
    if address.startswith('1') and 26 <= len(address) <= 35:
        return True
    if address.startswith('3') and 26 <= len(address) <= 35:
        return True
    if address.startswith('bc1') and 14 <= len(address) <= 74:
        return True
    
    return False

# Export main functions for backward compatibility
__all__ = [
    'ADDRESS_FORMATS',
    'generate_wepo_address',
    'validate_wepo_address',
    'get_address_type',
    'is_quantum_address',
    'is_regular_address',
    'format_address_for_display',
    'standardize_address',
    'get_address_patterns',
    'addresses_equal',
    'validate_address_batch',
    'is_valid_wepo_address',
    'validate_btc_address'
]